rs/ledger_suite/tests/sm-tests/src/fee_collector.rs

```diff
@@ -0,0 +1,272 @@
+                )
+                .unwrap();
+
+                // Record the previous total_supply and make the transfer.
+                let total_supply_before = total_supply(&env, ledger_id);
+                transfer(&env, ledger_id, account_from, account_to, amount)
+                    .expect("Unable to perform transfer");
+
+                // If the fee was burned then the total_supply after the
+                // transfer should be the one before plus the (burned) FEE.
+                assert_eq!(
+                    total_supply_before,
+                    total_supply(&env, ledger_id) + FEE,
+                    "Total supply should have been decreased of the (burned) fee {}",
+                    FEE
+                );
+
+                // The fee collector must have collected no fee this time.
+                assert_eq!(
+                    FEE,
+                    balance_of(&env, ledger_id, fee_collector),
+                    "The fee_collector should have collected the fee"
+                );
+
+                Ok(())
+            },
+        )
+        .unwrap();
+}
+
+pub fn test_fee_collector_blocks<T>(ledger_wasm: Vec<u8>, encode_init_args: fn(InitArgs) -> T)
+where
+    T: CandidType,
+{
+    fn value_as_u64(value: icrc_ledger_types::icrc::generic_value::Value) -> u64 {
+        use icrc_ledger_types::icrc::generic_value::Value;
+        match value {
+            Value::Nat64(n) => n,
+            Value::Nat(n) => n.0.to_u64().expect("block index should fit into u64"),
+            Value::Int(int) => int.0.to_u64().expect("block index should fit into u64"),
+            value => panic!("Expected a numeric value but found {:?}", value),
+        }
+    }
+
+    fn value_as_account(value: icrc_ledger_types::icrc::generic_value::Value) -> Account {
+        use icrc_ledger_types::icrc::generic_value::Value;
+
+        match value {
+            Value::Array(array) => match &array[..] {
+                [Value::Blob(principal_bytes)] => Account {
+                    owner: Principal::try_from(principal_bytes.as_ref())
+                        .expect("failed to parse account owner"),
+                    subaccount: None,
+                },
+                [Value::Blob(principal_bytes), Value::Blob(subaccount_bytes)] => Account {
+                    owner: Principal::try_from(principal_bytes.as_ref())
+                        .expect("failed to parse account owner"),
+                    subaccount: Some(
+                        Subaccount::try_from(subaccount_bytes.as_ref())
+                            .expect("failed to parse subaccount"),
+                    ),
+                },
+                _ => panic!("Unexpected account representation: {:?}", array),
+            },
+            value => panic!("Expected Value::Array but found {:?}", value),
+        }
+    }
+
+    fn fee_collector_from_block(
+        block: icrc_ledger_types::icrc::generic_value::Value,
```

Comment:
Why isn't `block` an immutable reference here? It seems we don't need ownership but we end up cloning it every time we call this method.

Reply:
Good question! I just moved this code from the parent module, but you're right that there's a bunch of unnecessary cloning going on. I changed all the helper functions to take immutable references and removed a bunch of calls to `clone()`.